### PR TITLE
Fixes #2099 -- Add GCP Stackdriver support to telegraf.

### LINKED ISSUE #2099: Add Google Cloud Monitoring input (Stackdriver)
### Proposal:

Google [has made Stackdriver generally available](https://cloudplatform.googleblog.com/2016/10/Google-Stackdriver-Generally-Available.html), and has a [metrics API](https://cloud.google.com/monitoring/docs/apis).

Please add support for fetching metrics from Google Cloud Monitoring, similar to [AWS CloudWatch input](https://github.com/influxdata/telegraf/tree/5d3850c44e048f5619a04c14373c3324c2c9c407/plugins/inputs/cloudwatch).

### Use case

Google Cloud Platform is a major cloud provider nowadays, and I'd prefer pulling all metrics into a single backend source (InfluxDB, Prometheus) and manage alerting from there, instead of having to manage various alerting sources (in this case, Google Stackdriver along with my main monitoring/alerting backend).

### PR DESCRIPTION
**Pull request by derrley**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [ ] Has appropriate unit tests. --  not yet.

This is my first PR -- looking to get some feedback on the direction of this. We've got it running in our own environment with mostly successful results, but I know others might have different environments. In the mean time, I'm going to be working on unit tests and beefing up the README. The example config should make the features fairly clear.

### CODE REVIEW

**Comment by danielnelson** [comment]
Thanks for the pull request @derrley, I'll try to take a closer look soon, probably during the 1.7 cycle.

**Comment by russorat** [comment]
@derrley thanks so much for this PR! Do you mind providing a little more information about your use case and how you use this in your environment?

**Comment by JanekLehr** [comment]
@russorat 
I worked with @derrley while he was developing this plugin. We run a TICK stack in Google Kubernetes Engine, and one of the telegraf instances scrapes Stackdriver metrics and sends them to InfluxDB. Currently the only Kapacitor alert that uses these metrics is one to alert when a pod container was OOMKilled, but it also gives us the ability to track latencies of our HTTP Load Balancer in GCP, and other GCP resource metrics.

By doing this we can have all our alerts in one place (kapacitor) and avoid defining, and paying for, alerts in Stackdriver

File: plugins/inputs/gcp_stackdriver/gcp_stackdriver.go [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+	return `
+  ## GCP Project
+  project = "projects/erudite-bloom-151019"
+
+  ## API rate limit. On a default project, it seems that a single user can make
+  ## ~14 requests per second. This might be configurable. Each API request can
+  ## fetch every time series for a single metric type, though -- this is plenty
+  ## fast for scraping all the builtin metric types (and even a handful of
+  ## custom ones) every 60s.
+  #rateLimit = 14
```
Comment (glinton) [inline plugins/inputs/gcp_stackdriver/gcp_stackdriver.go]: nitpick: add single space after `#`.

File: plugins/inputs/gcp_stackdriver/gcp_stackdriver.go [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+	return includeExcludeHelper(k, inc, exc)
+}
+
+// Generate a list of timeSeriesConfig structs by making a ListMetricDescriptors
+// API request and filtering the result against our configuration.
+func (s *GCPStackdriver) generatetimeSeriesConfs() ([]timeSeriesConf, error) {
+	ret := []timeSeriesConf{}
+	req := &monitoringpb.ListMetricDescriptorsRequest{Name: s.Project}
+	resp := s.client.ListMetricDescriptors(context.TODO(), req)
```
Comment (glinton) [inline plugins/inputs/gcp_stackdriver/gcp_stackdriver.go]: Consider using a context with a timeout or deadline.

File: plugins/inputs/gcp_stackdriver/gcp_stackdriver.go [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+				case metricpb.MetricDescriptor_BOOL:
+					field = p.Value.GetBoolValue()
+				case metricpb.MetricDescriptor_INT64:
+					field = p.Value.GetInt64Value()
+				case metricpb.MetricDescriptor_DOUBLE:
+					field = p.Value.GetDoubleValue()
+				case metricpb.MetricDescriptor_STRING:
+					field = p.Value.GetStringValue()
+				default:
+					// TODO: AddError here? Is it really an error?
```
Comment (glinton) [inline plugins/inputs/gcp_stackdriver/gcp_stackdriver.go]: Maybe either skip the metric, as you currently are, or default it to a chosen type.

File: plugins/inputs/gcp_stackdriver/gcp_stackdriver.go [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+		}
+
+		s.client = client
+	}
+
+	return nil
+}
+
+func includeExcludeHelper(key string, includes []string, excludes []string) bool {
+	if includes != nil {
```
Comment (glinton) [inline plugins/inputs/gcp_stackdriver/gcp_stackdriver.go]: maybe also check if `includes` and `excludes` isn't a 0 length slice

**Comment by glinton** [comment]
@derrley Do you still have time available, and are you interested in continuing this?Thanks

**Comment by danielnelson** [comment]
Latest work is on #4977

**Pull request closed without merging**